src/base/static/libs/maps/mapboxgl-provider.js

```diff
@@ -149,53 +336,70 @@ export default (container, options) => {
       };
     },
 
-    createGeoJSONLayer: options => {
-      map.addSource(options.id, {
+    getSource: sourceId => {
+      return map.getSource(sourceId);
+    },
+
+    createGeoJSONLayer: ({ id, url, rules }) => {
+      map.addSource(id, {
         type: "geojson",
-        data: options.url,
+        data: url,
       });
 
-      return options.rules.map((styleRule, i) => {
-        styleRule.id = [options.id, "_", i].join("");
-        styleRule.source = options.id;
+      return rules.map((styleRule, i) => ({
+        ...styleRule,
+        id: `${id}_${i}`,
+        source: id,
+      }));
+    },
 
-        return styleRule;
+    createPlaceLayer: ({ id, rules, geoJSON }) => {
+      map.addSource(id, {
+        type: "geojson",
+        data: geoJSON,
       });
+
+      // We need to add a filter on location_type for place layers. We append
+      // that filter programatically here to save repetition in the config.
+      return Object.entries(rules)
+        .map(([locationType, styleRules]) => {
+          const locationTypeFilter = [
+            "==",
+            ["get", constants.LOCATION_TYPE_PROPERTY_NAME],
+            locationType,
+          ];
+          return styleRules.map((styleRule, i) => ({
+            ...styleRule,
+            filter: appendFilters(styleRule.filter, locationTypeFilter),
+            source: id,
+            id: `${locationType}_${i}`,
+          }));
+        })
+        .reduce((flat, toFlatten) => {
+          return flat.concat(toFlatten);
+        }, []);
     },
 
     addLayer: layerConfig => {
       map.addLayer(layerConfig);
     },
 
-    addVectorLayerGroup: layerStyles => {
-      layerStyles.forEach(layerStyle => {
-        map.addLayer(layerStyle);
+    addVectorLayerGroup: layerConfigs => {
+      layerConfigs.forEach(layerConfig => {
+        map.addLayer(layerConfig);
       });
     },
 
-    addGeoJSONLayer: (layerStyles, geometryType) => {
-      // TODO: Can we detect GeoJSON geometry types automatically instead?
-      // What about datasets with multiple geometry types?
-      // It's possible to obtain the geometry type in filter expressions, e.g.:
-      // ["==", "$type", "Point"]
-
-      if (geometryType === "Point") {
-        layerStyles.forEach(layerStyle => {
-          layerStyle.layout["icon-allow-overlap"] = true;
-          layerStyle.type = "symbol";
-          map.addLayer(layerStyle);
-        });
-      } else if (geometryType === "Polygon") {
-        layerStyles.forEach(layerStyle => {
-          layerStyle.type = "fill";
-          map.addLayer(layerStyle);
-        });
-      } else if (geometryType === "LineString") {
-        layerStyles.forEach(layerStyle => {
-          layerStyle.type = "line";
-          map.addLayer(layerStyle);
-        });
-      }
+    addGeoJSONLayer: layerConfigs => {
+      layerToSymbol(layerConfigs).forEach(layerConfig => {
```

Comment:
as mentioned above, I think this would be easier to understand if it was written:

`layerConfigs.map(layerToSymbol).forEach(symbol => {`

Do you agree?

Also, I'm still not even sure what is the difference between a "layer config" vs a "layer" vs a "geoJSON layer" etc. We have so many different data structures floating around, and many representing the same thing, we should really straighten this out at some point. I am happy to discuss and help out if you think it's a priority.

Reply:
Also, I imagine that the `layerRepresentation`s are what the map layer duck will manage, whether that's a cache of Leaflet layer objects or an array of MapboxGL layer style definitions.

Reply:
As discussed, we are going to need a "first class" representation of our state, but we are not quite sure what that is yet. Currently, we have 3 representations:

 1. the POJO returned from `createXLayer`, which we store in `this.layers`. This is the closest thing we have to a "first class" representation because it is stored in our app state.
 2. The raw "layer config" POJO, generated after parsing the YAML config
 3. The Map Provider representation returned from our `layerToSymbol/Line/Fill` functions, which is passed to the MapBox `map.addLayer` function. 

We are going to punt on figuring this out for now, until the port is finished.

In the meantime, it could be useful to establish a naming convention like the one suggested above. But we'll need 3 names instead of 2. I think we can name 1) Layer, 2) LayerConfig and 3) LayerRepresentation. Does that seem reasonable?

Reply:
I think it would be reasonable in this PR to iron out the `layer`/`layerRepresentation` distinction such that we have only two representations of our layers: the config version, and the "processed" layer representation that is provider-specific.

So a call to `createLayer` would take in a `layerConfig` from the `config.yml` and return a `layer`. The `layer` is cached in `this.layers`, and is added to the map when needed with `map.addLayer()`.

Reply:
A wrinkle here is the fact that a single config-derived layer might actually consist of multiple internal layers in Mapbox's terms. Technically we create a layer for each filter condition and each geometry type for a given config layer.

I think that our first-class layer representation could in fact be an array of layer style definitions (each one itself a layer in Mapbox terms), and `addLayer` could assume it will receive an array and call the internal `map.addLayer()` once for each element in the array.

Reply:
>I think that our first-class layer representation could in fact be an array of layer style definitions (each one itself a layer in Mapbox terms)

I don't think that is a good idea because:
 1) it complicates our layer schema by making it a collection, which is only needed because that's how our Mapbox Provider wants it
 2) it couples our layer to the Map Provider representation. This adds complexity as well, because then we wouldn't be able to make any assumptions about the schema of our layer. eg: How can we ensure that we can access `layer.source` if we can't make any assumptions about whether the layer has a `source` attribute?

I think it would be simpler to have our first-class layer representation upstream from the map provider representation. Whether the first-class representation is the layer config itself, or a processed version of the config, isn't a big deal right now.

But if you really wanna give it a shot, then I'll trust your judgement :-)

Reply:
> it couples our layer to the Map Provider representation

That's a fair point. 

> Whether the first-class representation is the layer config itself, or a processed version of the config, isn't a big deal right now

Well the processed version of the config essentially is the collection of layers referred to above, so I think we should store the raw layer config upstream if we want to keep layers decoupled from the Provider.

We'll need to maintain some bookkeeping to link a config layer (i.e. the thing that users will think of as a map layer) to its constituent internal Mapbox layers. But that's probably straightforward.

I suppose that's what the map layer duck should do-- store the config representations of layers, and do the bookkeeping for layer grouping, visibility, and z-ordering.

Reply:
>Well the "processed" version of the config essentially is the collection of layers referred to above, so I think we should store the raw layer config upstream if we want to keep layers decoupled from the Provider.

I was thinking about the POJO returned from `createGeoJSONLayer` vs the multiple POJO's passed to `map.addLayer` in `addGeoJSONLayer`. It doesn't seem like the "processed" config has to be the map provider representation - we can still do some config processing to get the first-class representation without processing it all the way to a map provider representation. For example, there may be some YAML (or eventually db-specific structures) that we would want to represent differently in app state. But we should stick with the raw layer config as much as possible.


>I suppose that's what the map layer duck should do-- store the config representations of layers, and do the bookkeeping for layer grouping, visibility, and z-ordering.

I don't know what all of that entails, but if it can be done through managing an array of Layer objects, then yes. Anything to do with the state of the Layers should be managed by the Layer duck.